[17.0][MIG] sale_timesheet_invoice_description

MarioLM-23 opened the pull request:
Standard migration with changes of old references

carlosdauden committed:
[9.0][NEW] sale_timesheet_invoice_description: New module (#154)

[ADD] sale_timesheet_invoice_description: New module

oca-transbot committed:
OCA Transbot updated translations from Transifex

sbidoul committed:
[FIX] remove en.po that was erroneously created by transbot

oca-transbot committed:
OCA Transbot updated translations from Transifex

pedrobaeza committed:
[MIG] Make modules uninstallable

pedrobaeza committed:
[MIG] Rename manifest files

oca-transbot committed:
OCA Transbot updated translations from Transifex

pedrobaeza committed:
[MIG] sale_timesheet_invoice_description: Migration to 10.0

oca-transbot committed:
OCA Transbot updated translations from Transifex

astirpe committed:
[11.0][MIG] sale_timesheet_invoice_description: port to V11

oca-transbot committed:
OCA Transbot updated translations from Transifex

oca-travis committed:
[UPD] Update sale_timesheet_invoice_description.pot

pedrobaeza committed:
[FIX] sale_timesheet_invoice_description: Load en_US language in tests

pedrobaeza committed:
[FIX] sale_timesheet_invoice_description: Avoid new constraint

There's a constraint that limits the quantity over received one. Although it seems invalid
on services, we bypass it now with this.

rschnapka committed:
Translated using Weblate (German)

Currently translated at 100,0% (10 of 10 strings)

Translation: account-invoicing-11.0/account-invoicing-11.0-sale_timesheet_invoice_description
Translate-URL: https://translation.odoo-community.org/projects/account-invoicing-11-0/account-invoicing-11-0-sale_timesheet_invoice_description/de/

ernestotejeda committed:
[MIG] sale_timesheet_invoice_description: Migration to 12.0

ernestotejeda committed:
[MIG] sale_timesheet_invoice_description: Fix E261 flake8 error

ernestotejeda committed:
[MIG] sale_timesheet_invoice_description: Improve code

Keep the previous behavior showing only date
Remove order parameter in sear method

oca-travis committed:
[UPD] Update sale_timesheet_invoice_description.pot

oca-transbot committed:
Update translation files

Updated by Update PO files to match POT (msgmerge) hook in Weblate.

OCA-git-bot committed:
[ADD] icon.png

josepmy committed:
Translated using Weblate (Spanish)

Currently translated at 100.0% (10 of 10 strings)

Translation: account-invoicing-12.0/account-invoicing-12.0-sale_timesheet_invoice_description
Translate-URL: https://translation.odoo-community.org/projects/account-invoicing-12-0/account-invoicing-12-0-sale_timesheet_invoice_description/es/

OCA-git-bot committed:
[UPD] README.rst

marylla committed:
Translated using Weblate (German)

Currently translated at 100.0% (10 of 10 strings)

Translation: account-invoicing-12.0/account-invoicing-12.0-sale_timesheet_invoice_description
Translate-URL: https://translation.odoo-community.org/projects/account-invoicing-12-0/account-invoicing-12-0-sale_timesheet_invoice_description/de/

Commit:
Translated using Weblate (Portuguese (Brazil))

Currently translated at 100.0% (10 of 10 strings)

Translation: account-invoicing-12.0/account-invoicing-12.0-sale_timesheet_invoice_description
Translate-URL: https://translation.odoo-community.org/projects/account-invoicing-12-0/account-invoicing-12-0-sale_timesheet_invoice_description/pt_BR/

Commit:
[IMP] sale_timesheet_invoice_description: black, isort

Commit:
[MIG] sale_timesheet_invoice_description: Migration to 13.0

oca-travis committed:
[UPD] Update sale_timesheet_invoice_description.pot

oca-transbot committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: account-invoicing-13.0/account-invoicing-13.0-sale_timesheet_invoice_description
Translate-URL: https://translation.odoo-community.org/projects/account-invoicing-13-0/account-invoicing-13-0-sale_timesheet_invoice_description/

Commit:
Translated using Weblate (Portuguese (Brazil))

Currently translated at 100.0% (10 of 10 strings)

Translation: account-invoicing-13.0/account-invoicing-13.0-sale_timesheet_invoice_description
Translate-URL: https://translation.odoo-community.org/projects/account-invoicing-13-0/account-invoicing-13-0-sale_timesheet_invoice_description/pt_BR/

oca-transbot committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: account-invoicing-13.0/account-invoicing-13.0-sale_timesheet_invoice_description
Translate-URL: https://translation.odoo-community.org/projects/account-invoicing-13-0/account-invoicing-13-0-sale_timesheet_invoice_description/

bosd committed:
Translated using Weblate (Dutch)

Currently translated at 90.0% (9 of 10 strings)

Translation: account-invoicing-13.0/account-invoicing-13.0-sale_timesheet_invoice_description
Translate-URL: https://translation.odoo-community.org/projects/account-invoicing-13-0/account-invoicing-13-0-sale_timesheet_invoice_description/nl/

anasuarez1 committed:
Translated using Weblate (Spanish)

Currently translated at 100.0% (10 of 10 strings)

Translation: account-invoicing-13.0/account-invoicing-13.0-sale_timesheet_invoice_description
Translate-URL: https://translation.odoo-community.org/projects/account-invoicing-13-0/account-invoicing-13-0-sale_timesheet_invoice_description/es/

clementmbr committed:
[IMP] sale_timesheet_invoice_description: black, isort, prettier

clementmbr committed:
[MIG] sale_timesheet_invoice_description: Migration to 14.0

- **Changes on tests :** now the timesheet needs to be linked to a
project task and this task is linked to the sale.order.line.

- **Changes on sale.py :** we cannot anymore find the timesheet related
to our future invoice line checking its create_date as v14 allows
to invoice only timesheets between two dates.
So I decided to refactor completely the way to match an invoice line
and its timesheet adding a `_link_timesheets_to_invoice_line` method
inspired by the native `_link_timesheets_to_invoice`

clementmbr committed:
[IMP] option to split order.line in one invoice line for each timesheet

clementmbr committed:
[UPD] Readme

oca-travis committed:
[UPD] Update sale_timesheet_invoice_description.pot

ibuioli committed:
Translated using Weblate (Spanish (Argentina))

Currently translated at 100.0% (10 of 10 strings)

Translation: account-invoicing-14.0/account-invoicing-14.0-sale_timesheet_invoice_description
Translate-URL: https://translation.odoo-community.org/projects/account-invoicing-14-0/account-invoicing-14-0-sale_timesheet_invoice_description/es_AR/

Commit:
[IMP] sale_timesheet_invoice_description: black, isort, prettier

Commit:
[MIG] sale_timesheet_invoice_description: Migration to 15.0

Commit:
[UPD] Update sale_timesheet_invoice_description.pot

OCA-git-bot committed:
[UPD] README.rst

dsolanki-initos committed:
[IMP] sale_timesheet_invoice_description: pre-commit stuff

dsolanki-initos committed:
[MIG] sale_timesheet_invoice_description: Migration to 16.0

* adapted code to version 16.0 changes; for details see below
* added contributors
* updated i18n: sale_timesheet_invoice_description.pot, de.po
* run `pre-commit`

Linking timesheets to invoice lines: Respecting both the specified (in
context) start and end time, the same way as it is done for linking them
to the invoice itself.

Reworked preventing unbalanced invoice: Invoices must not and cannot be
unbalanced. This is now (in 16.0) done automatically. I.e., the
intermediate invoices, created while splitting lines, will no longer be
possibly unbalanced.

Commit:
[UPD] Update sale_timesheet_invoice_description.pot

weblate committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: account-invoicing-16.0/account-invoicing-16.0-sale_timesheet_invoice_description
Translate-URL: https://translation.odoo-community.org/projects/account-invoicing-16-0/account-invoicing-16-0-sale_timesheet_invoice_description/

Ivorra78 committed:
Translated using Weblate (Spanish)

Currently translated at 100.0% (17 of 17 strings)

Translation: account-invoicing-16.0/account-invoicing-16.0-sale_timesheet_invoice_description
Translate-URL: https://translation.odoo-community.org/projects/account-invoicing-16-0/account-invoicing-16-0-sale_timesheet_invoice_description/es/

adrianojprado committed:
Translated using Weblate (Portuguese (Brazil))

Currently translated at 100.0% (17 of 17 strings)

Translation: account-invoicing-16.0/account-invoicing-16.0-sale_timesheet_invoice_description
Translate-URL: https://translation.odoo-community.org/projects/account-invoicing-16-0/account-invoicing-16-0-sale_timesheet_invoice_description/pt_BR/

francesco-ooops committed:
Translated using Weblate (Italian)

Currently translated at 82.3% (14 of 17 strings)

Translation: account-invoicing-16.0/account-invoicing-16.0-sale_timesheet_invoice_description
Translate-URL: https://translation.odoo-community.org/projects/account-invoicing-16-0/account-invoicing-16-0-sale_timesheet_invoice_description/it/

mymage committed:
Translated using Weblate (Italian)

Currently translated at 100.0% (17 of 17 strings)

Translation: account-invoicing-16.0/account-invoicing-16.0-sale_timesheet_invoice_description
Translate-URL: https://translation.odoo-community.org/projects/account-invoicing-16-0/account-invoicing-16-0-sale_timesheet_invoice_description/it/

jakobkrabbe committed:
Translated using Weblate (Swedish)

Currently translated at 100.0% (17 of 17 strings)

Translation: account-invoicing-16.0/account-invoicing-16.0-sale_timesheet_invoice_description
Translate-URL: https://translation.odoo-community.org/projects/account-invoicing-16-0/account-invoicing-16-0-sale_timesheet_invoice_description/sv/

Commit:
[IMP] sale_timesheet_invoice_description: pre-commit auto fixes

bosd commented:
Is it me or is there more going on in your first commit then just isort,prettier, black and pre-commit?

bosd commented on sale_timesheet_invoice_description/__manifest__.py:
```diff
@@ -0,0 +1,17 @@
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+{
+    "name": "Timesheet details invoice",
```
```suggestion
    "name": "Timesheet details invoice description",
```

bosd commented on sale_timesheet_invoice_description/readme/USAGE.md:
```diff
@@ -0,0 +1,22 @@
+    (timesheet line) and select a task related to the Sale Order's
+    project (and to your specific Sale Order's line). Remember to add
+    the time spent.
+6.  Go to the Sale Order and create its invoice (clic on 'Create
```
```suggestion
6.  Go to the Sale Order and create its invoice (click on 'Create
```

bosd commented on sale_timesheet_invoice_description/readme/USAGE.md:
```diff
@@ -0,0 +1,22 @@
+    - **Unit of Measure** \> 'Hours' (or any other time unit)
+2.  Go to *Sales \> Orders \> Orders*, select a Sale Order or create a
+    new one, add a product (service) under 'Order Lines' tab.
+3.  Go to 'Other Info' tab (wihthin the same Sale Order) and:
```
```suggestion
3.  Go to 'Other Info' tab (within the same Sale Order) and:
```

bosd commented on sale_timesheet_invoice_description/tests/test_sale_timesheet_description.py:
```diff
@@ -0,0 +1,250 @@
+
+from datetime import datetime
+
+from odoo.tests import common
```
```suggestion
from odoo.tests import common
from odoo.addons.base.tests.common import DISABLED_MAIL_CONTEXT
```

bosd commented on sale_timesheet_invoice_description/tests/test_sale_timesheet_description.py:
```diff
@@ -0,0 +1,250 @@
+    @classmethod
+    def setUpClass(cls):
+        super().setUpClass()
+        # Make sure user is in English
```
```suggestion
cls.env = cls.env(context=dict(cls.env.context, **DISABLED_MAIL_CONTEXT))
        # Make sure user is in English
```

bosd reviewed (commented):
Some minor improvements

MarioLM-23 commented:
> Is it me or is there more going on in your first commit then just isort,prettier, black and pre-commit?

Yes sorry I changed the version in manifest, in the test file the super setUpClass and added contributors

MarioLM-23 commented:
@bosd 

I did the changes and improve the reference in xpath of configuration view, could you check it and let me know if I have done it correctly?
Thanks for the review!

bosd commented:
I've encountered an error, while testing on the runboat the default description feature.
Steps to reproduce.
On a SO set the Description to empty.
![image](https://github.com/OCA/account-invoicing/assets/11499387/25e9ff9f-cb36-40bf-a02f-787e6d323dca)

Then create an invoice:

Error occurs:
```
Traceback (most recent call last):
  File "/opt/odoo/odoo/http.py", line 1768, in _serve_db
    return service_model.retrying(self._serve_ir_http, self.env)
  File "/opt/odoo/odoo/service/model.py", line 133, in retrying
    result = func()
  File "/opt/odoo/odoo/http.py", line 1795, in _serve_ir_http
    response = self.dispatcher.dispatch(rule.endpoint, args)
  File "/opt/odoo/odoo/http.py", line 1999, in dispatch
    result = self.request.registry['ir.http']._dispatch(endpoint)
  File "/opt/odoo/odoo/addons/base/models/ir_http.py", line 222, in _dispatch
    result = endpoint(**request.params)
  File "/opt/odoo/odoo/http.py", line 725, in route_wrapper
    result = endpoint(self, *args, **params_ok)
  File "/opt/odoo/addons/web/controllers/dataset.py", line 28, in call_button
    action = self._call_kw(model, method, args, kwargs)
  File "/opt/odoo/addons/web/controllers/dataset.py", line 20, in _call_kw
    return call_kw(request.env[model], method, args, kwargs)
  File "/opt/odoo/odoo/api.py", line 468, in call_kw
    result = _call_kw_multi(method, model, args, kwargs)
  File "/opt/odoo/odoo/api.py", line 453, in _call_kw_multi
    result = method(recs, *args, **kwargs)
  File "/opt/odoo/addons/sale/wizard/sale_make_invoice_advance.py", line 180, in create_invoices
    invoices = self._create_invoices(self.sale_order_ids)
  File "/opt/odoo/addons/sale_timesheet/wizard/sale_make_invoice_advance.py", line 46, in _create_invoices
    return sale_orders.with_context(
  File "/mnt/data/odoo-addons-dir/sale_timesheet_invoice_description/models/sale.py", line 131, in _create_invoices
    desc_list = self._get_timesheet_description_list(ts_ids, desc_rule)
  File "/mnt/data/odoo-addons-dir/sale_timesheet_invoice_description/models/sale.py", line 41, in _get_timesheet_description_list
    details = self._get_timesheet_details(timesheet_id, desc_rule)
  File "/mnt/data/odoo-addons-dir/sale_timesheet_invoice_description/models/sale.py", line 29, in _get_timesheet_details
    if desc_rule[0] == "1":
TypeError: 'bool' object is not subscriptable

The above server error caused the following client error:
RPC_ERROR: Odoo Server Error
    at makeErrorFromResponse (http://oca-account-invoicing-17-0-pr1713-fd442d552117.runboat.odoo-community.org/web/assets/0ab7b66/web.assets_web.min.js:2870:163)
    at XMLHttpRequest.<anonymous> (http://oca-account-invoicing-17-0-pr1713-fd442d552117.runboat.odoo-community.org/web/assets/0ab7b66/web.assets_web.min.js:2874:13)
```

I was expecting that the default method would be used insted. Which s configured in configuration--> settings.


EDIT: My Bad, for using incorrect workflow.
When creating a a new SO. The correct description is applied. (the one defined under config/settings).
According to this error it is not allowed to have the empty value assigned.

bosd commented:
My last test actually leads me to reproduce another error.
After installation of this module. The default Timesheet description is on the empty value.
When creating a new invoice. This empty value is assigned. Leading to afforementioned error.

Proposal:
Remove the Empy value from the Invoice methods.
By default assign the value `[000] none` 

![image](https://github.com/OCA/account-invoicing/assets/11499387/1d0ae38e-ca07-4ec8-b855-b214402177ac)

MarioLM-23 commented:
@bosd Hello and thank you for checking those errors
The default value was already set to none, to solve the error of setting the description to empty I set the field in the views as required, with this it is solved?
I first tried to set the field in the model as required but it was only applied in the order view

bosd commented:
@mariobinhex Thanks for your patience and working on this bug.

I just discovered that this bug was already present in earlier versions.
So preferrebly the changes should be placed in a seperate commit for easy backport :hugs: 

The required flag works as inteded. It is no longer possible to assign the empty `False` value.
But, right after installation of this module. In configuration settings the value is still `False`.
We should set a default in python code. (just like on SO)
add
`default="000"`

at
https://github.com/OCA/account-invoicing/pull/1713/files#diff-3f0a25bc7aaab05ba3a84450130c78677a449cd2abbc4644e07786804d690a76R9-R13

If this module is installed in a DB with exsisting SO's. The error could still occur. (Because the value is False).
Or possibly when the SO is created via non UI.
For robustness we should adopt the code to allow for a `False` situation, I think changing this
https://github.com/OCA/account-invoicing/pull/1713/files#diff-26928443132d5ac74840f1c074ae427e24bdd060867bab3437730a4cc3e7c763R27-R35
to:
```
    def _get_timesheet_details(self, timesheet, desc_rule):
        details = []
        if desc_rule is False:
        return details
        if desc_rule[0] == "1":
            details.append(fields.Date.to_string(timesheet.date))
        if desc_rule[1] == "1":
            details.append(f"{timesheet.unit_amount} {timesheet.product_uom_id.name}")
        if desc_rule[2] == "1":
            details.append(timesheet.name)
        return details

```
Should do the trick.

MarioLM-23 commented:
@bosd Done! Please check my last commit when you can and tell me if any further changes are necessary

bosd commented:
@mariobinhex Thanks, please move the required flags to the last commit.
(For easy cherry picking, backports)

Please change the [commit message according to OCA guidelines](https://github.com/OCA/odoo-community.org/blob/master/website/Contribution/CONTRIBUTING.rst#commit-message)
`[IMP] sale_timesheet_invoice_description: Handle default description`
Or a better description 😄

Commit:
[MIG] sale_timesheet_invoice_description: Migration to 17.0

Commit:
[FIX] sale_timesheet_invoice_description: Fix bug TID
Fix bug of setting Timesheet Invoice Description field as false by default or in the configuration and sales view
Improve sale model allowing for a False situation of Timesheet Invoice Description

MarioLM-23 commented:
@bosd Done, thanks for your patience looking for my mistakes and improving the code

bosd reviewed (approved):
LGTM ✨  Thanks for contributing

Dranyel-Bosd reviewed (approved):
LGTM :+1:  :sparkler:

remi-filament reviewed (approved):
LGTM thank you for your work @MarioLM-23

pedrobaeza commented:
Please squash last 2 commits into one, as both belongs to the migration itself.

bosd commented:
> Please squash last 2 commits into one, as both belongs to the migration itself.

I've actually requested to have the fix of the bug I've discovered to be in a seperate commit. 
(Bug was present in V15+, likely in 14 as well)
For easier cherry picking / backporting. (Which is now in the last commit)
@pedrobaeza Is that ok to you?

pedrobaeza commented:
That commit contains changes in the README, which doesn't seem correct.

andreschenkels commented:
@pedrobaeza @MarioLM-23 is it possible to resolve the questions so it can be merged? It's done but never merged.

pedrobaeza commented on sale_timesheet_invoice_description/views/sale_view.xml:
```diff
@@ -8,7 +8,7 @@
         <field name="arch" type="xml">
             <field name="show_update_fpos" position="after">
-                <field name="timesheet_invoice_description" />
+                <field name="timesheet_invoice_description" required="1" />
```
Why forcing this?

victoralmau commented on sale_timesheet_invoice_description/tests/test_sale_timesheet_description.py:
```diff
@@ -0,0 +1,253 @@
+from odoo.tests import common
+from odoo.tools.float_utils import float_compare
+
+from odoo.addons.base.tests.common import DISABLED_MAIL_CONTEXT
+
+
+class TestSaleTimesheetDescription(common.TransactionCase):
+    @classmethod
+    def setUpClass(cls):
+        super().setUpClass()
+        cls.env = cls.env(context=dict(cls.env.context, **DISABLED_MAIL_CONTEXT))
```
```suggestion
from odoo.addons.base.tests.common import BaseCommon


class TestSaleTimesheetDescription(BaseCommon):
    @classmethod
    def setUpClass(cls):
        super().setUpClass()
```

victoralmau commented:
Superseded by https://github.com/OCA/account-invoicing/pull/1985
